mmcv/cnn/bricks/transformer.py

```diff
@@ -0,0 +1,433 @@
+import copy
+
+import torch.nn as nn
+
+from mmcv import ConfigDict
+from mmcv.cnn import Linear, build_activation_layer, build_norm_layer
+from mmcv.utils import build_from_cfg
+from .registry import (ATTENTION, POSITIONAL_ENCODING, TRANSFORMERCODER,
+                       TRANSFORMERLAYER)
+
+
+def build_positional_encoding(cfg, default_args=None):
+    """Builder for Position Encoding."""
+    return build_from_cfg(cfg, POSITIONAL_ENCODING, default_args)
+
+
+def build_attention(cfg, default_args=None):
+    """Builder for attention."""
+    return build_from_cfg(cfg, ATTENTION, default_args)
+
+
+def build_transformerlayer(cfg, default_args=None):
+    """Builder for transformer layer."""
+    return build_from_cfg(cfg, TRANSFORMERLAYER, default_args)
+
+
+def build_transformercoder(cfg, default_args=None):
+    """Builder for transformer coder."""
+    return build_from_cfg(cfg, TRANSFORMERCODER, default_args)
+
+
+class MultiheadAttention(nn.Module):
+    """A warpper for torch.nn.MultiheadAttention.
+
+    This module implements MultiheadAttention with residual connection,
+    and positional encoding used in DETR is also passed as input.
+
+    Args:
+        embed_dims (int): The embedding dimension.
+        num_heads (int): Parallel attention heads. Same as
+            `nn.MultiheadAttention`.
+        dropout (float): A Dropout layer on attn_output_weights. Default: 0.0.
+    """
+
+    def __init__(self, embed_dims, num_heads, dropout=0.0):
+        super(MultiheadAttention, self).__init__()
+        assert embed_dims % num_heads == 0, 'embed_dims must be ' \
+            f'divisible by num_heads. got {embed_dims} and {num_heads}.'
+        self.embed_dims = embed_dims
+        self.num_heads = num_heads
+        self.dropout = dropout
+        self.attn = nn.MultiheadAttention(embed_dims, num_heads, dropout)
+        self.dropout = nn.Dropout(dropout)
+
+    def forward(self,
+                query,
+                key=None,
+                value=None,
+                residual=None,
+                query_pos=None,
+                key_pos=None,
+                attn_mask=None,
+                key_padding_mask=None,
+                **kwargs):
+        """Forward function for `MultiheadAttention`.
+
+        Args:
+            query (Tensor): The input query with shape [num_query, bs,
+                embed_dims]. Same in `nn.MultiheadAttention.forward`.
+            key (Tensor): The key tensor with shape [num_key, bs,
+                embed_dims]. Same in `nn.MultiheadAttention.forward`.
+                Default None. If None, the `query` will be used.
```

Comment:
You may change the order of these two sentences:
If None, the `query` will be used. Defaults to None.

Reply:
In Sphinx,  \`x\` denotes *italic texts* while \`\`x \`\` will render texts to `in-line code`.